src/Utils.cpp

```diff
@@ -499,6 +499,82 @@ CString CUtils::FormatTime(time_t t, const CString& sFormat,
     return s;
 }
 
+/* Supports additional format specifiers for formatting sub-second values:
+ *
+ * - %L - millisecond, 3 digits (ruby extension)
+ * - %N - sub-second fraction (ruby extension)
+ *   - %3N - millisecond
+ *   - %6N - microsecond
+ *   - %9N - nanosecond (default if no digit specified)
+ *
+ * However, note that timeval only supports microsecond precision
+ * (thus, formatting with higher-than-microsecond precision will
+ * always result in trailing zeroes), and IRC server-time is specified
+ * in millisecond precision (thus formatting received timestamps with
+ * higher-than-millisecond precision will always result in trailing
+ * zeroes).
+ */
+CString CUtils::FormatTime(timeval tv, const CString& sFormat,
+                           const CString& sTimezone) {
+    // Parse additional format specifiers before passing them to
+    // strftime, since the way strftime treats unknown format
+    // specifiers is undefined.
+    CString sFormat2;
+
+    // Make sure %% is parsed correctly, i.e. %%L is passed through to
+    // strftime to become %L, and not 123.
+    bool bInFormat = false;
+    int iDigits;
+    CString::size_type pLastCopied = 0, pFormatStart;
+
+    for (CString::size_type i = 0; i < sFormat.length(); i++) {
+        if (!bInFormat) {
+            if (sFormat[i] == '%') {
+                pFormatStart = i;
+                bInFormat = true;
+                iDigits = 9;
+            }
+        } else {
+            switch (sFormat[i]) {
+                case '0': case '1': case '2': case '3': case '4':
+                case '5': case '6': case '7': case '8': case '9':
+                    iDigits = sFormat[i] - '0';
+                    break;
+                case 'L':
+                    iDigits = 3;
+                    // fall-through
+                case 'N': {
+                    int iVal = tv.tv_usec;
+                    int iDigitDelta = iDigits - 6; // tv_user is in 10^-6 seconds
+                    for (; iDigitDelta > 0; iDigitDelta--)
+                        iVal *= 10;
```

Comment:
This is likely to trigger signed overflow, which is undefined behavior in C++

Reply:
The possible range of valid values cannot result in an overflow. It's only possible if the timeval was invalid.